### PR TITLE
rules: allow creating a batch of ingress/egress rules

### LINKED ISSUE #194: security group rule mapping: allow multiple rules
Having to create / manage each rule separately can be tedious. This provider should allow to set rules in the same fashion than the cloudstack one.

cf. https://www.terraform.io/docs/providers/cloudstack/r/security_group_rule.html

### PR DESCRIPTION
**Pull request by greut**
Fixes #194 

```hcl
resource "exoscale_security_group_rules" "rules" {
  security_group_id = "${exoscale_security_group.sg.id}"

  ingress {
    protocol = "TCP"
    cidr_list = ["10.0.0.0/24", "::/0"]
    ports = ["22", "8000-8888"]
  }
}
```

**Problems**

<details>
<summary>subtle deletion</summary>
Assuming the above definition, four rules will be created. Now, one of them, e.g. `(::/0, 22)` is removed from the UI. When the _read_ operation occurs it sees that one of the four rules is missing although, that state cannot be properly represented with with `cidr_list` and `ports`. I can only represent the cross product of them.

Potential solutions: if any factor, e.g. `::/0` doesn't have _|ports|_ representations, then it should be removed from the list. Although we shall keep the ID in the _ids_ list. Deleting one rule from via the API will force to recreate the whole set of rules. Removing all the remaining ones and recreating them all.
</details>

**TODO**

- [x] doc
- [x] create / delete
- [x] read
- [x] update
- [x] subtle deletion
- [ ] clean code

### CODE REVIEW

File: exoscale/validation.go [hunk 1]
```diff
@@ -47,3 +48,43 @@ func ValidateIPv6String(i interface{}, k string) (s []string, es []error) {
...
+
+	ports := strings.Split(value, "-")
+	if len(ports) > 2 {
+		es = append(es, fmt.Errorf("expected a range of at most two values, got %d", len(ports)))
+		return
+	}
+
+	ps := make([]int, len(ports))
+	for i, port := range ports {
+		p, err := strconv.Atoi(port)
```
Comment (mcorbin) [inline exoscale/validation.go]: `strconv.ParseUint`

**Comment by greut** [comment]
d7fce771032f9241c41a32fadb8596a496df355d43e9142786334e6173786602 [terraform-provider-exoscale_SNAPSHOT-ec68336f5e0cc2f8a91c27b60e4f90e01a0b24e
9_linux_amd64.tar.gz](https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_SNAPSHOT-ec68336f5e0cc2f8a91c27b60e4f90e01a0b24e9_linux_amd64.tar.gz)

**Comment by mcorbin** [comment]
I have this Terraform configuration

```
resource "exoscale_security_group" "test1" {
  name = "testtf"
  description = "Default Security Group"
}

resource "exoscale_security_group_rules" "rules" {
  security_group_id = "${exoscale_security_group.test1.id}"
  ingress {
    protocol = "TCP"
    cidr_list = ["0.0.0.0/0"]
    ports = ["22"]
  }
  ingress {
    protocol = "ICMP"
    icmp_type = 8
    icmp_code = 0
    cidr_list = ["0.0.0.0/0"]
  }
}
```

The plan is:

```
  + exoscale_security_group.test1
      id:                                            <computed>
      description:                                   "Default Security Group"
      name:                                          "testtf"

  + exoscale_security_group_rules.rules
      id:                                            <computed>
      ingress.#:                                     "2"
      ingress.2724200827.cidr_list.#:                "1"
      ingress.2724200827.cidr_list.217649824:        "0.0.0.0/0"
      ingress.2724200827.description:                ""
      ingress.2724200827.icmp_code:                  ""
      ingress.2724200827.icmp_type:                  ""
      ingress.2724200827.ids.#:                      <computed>
      ingress.2724200827.ports.#:                    "1"
      ingress.2724200827.ports.2228573625:           "22"
      ingress.2724200827.protocol:                   "TCP"
      ingress.2724200827.user_security_group_list.#: "0"
      ingress.727548830.cidr_list.#:                 "1"
      ingress.727548830.cidr_list.217649824:         "0.0.0.0/0"
      ingress.727548830.description:                 ""
      ingress.727548830.icmp_code:                   "0"
      ingress.727548830.icmp_type:                   "8"
      ingress.727548830.ids.#:                       <computed>
      ingress.727548830.ports.#:                     "0"
      ingress.727548830.protocol:                    "ICMP"
      ingress.727548830.user_security_group_list.#:  "0"
      security_group:                                <computed>
      security_group_id:                             "${exoscale_security_group.test1.id}"


Plan: 2 to add, 0 to change, 0 to destroy.

```

But in the Interface, I have:

security

The ICMP rule has been replaced by an `ALL` rule.

If I do again a "terraform plan", I have: 

```
Terraform will perform the following actions:

  ~ exoscale_security_group_rules.rules
      ingress.2724200827.cidr_list.#:                "1" => "1"
      ingress.2724200827.cidr_list.217649824:        "0.0.0.0/0" => "0.0.0.0/0"
      ingress.2724200827.description:                "" => ""
      ingress.2724200827.icmp_code:                  "0" => "0"
      ingress.2724200827.icmp_type:                  "0" => "0"
      ingress.2724200827.ports.#:                    "1" => "1"
      ingress.2724200827.ports.2228573625:           "22" => "22"
      ingress.2724200827.protocol:                   "TCP" => "TCP"
      ingress.2724200827.user_security_group_list.#: "0" => "0"
      ingress.4142503364.cidr_list.#:                "1" => "0"
      ingress.4142503364.cidr_list.217649824:        "0.0.0.0/0" => ""
      ingress.4142503364.description:                "" => ""
      ingress.4142503364.icmp_code:                  "0" => "0"
      ingress.4142503364.icmp_type:                  "8" => "0"
      ingress.4142503364.ports.#:                    "1" => "0"
      ingress.4142503364.ports.798281000:            "0" => ""
      ingress.4142503364.protocol:                   "ALL" => ""
      ingress.4142503364.user_security_group_list.#: "0" => "0"
      ingress.727548830.cidr_list.#:                 "0" => "1"
      ingress.727548830.cidr_list.217649824:         "" => "0.0.0.0/0"
      ingress.727548830.description:                 "" => ""
      ingress.727548830.icmp_code:                   "" => "0"
      ingress.727548830.icmp_type:                   "" => "8"
      ingress.727548830.ids.#:                       "" => <computed>
      ingress.727548830.ports.#:                     "0" => "0"
      ingress.727548830.protocol:                    "" => "ICMP"
      ingress.727548830.user_security_group_list.#:  "0" => "0"


Plan: 0 to add, 1 to change, 0 to destroy.

```

Also, is it possible to separate the rules by a delimiter in the generated plan ?

**Comment by greut** [comment]
> The ICMP rule has been replaced by an ALL rule.

Good catch. That's what the `terraform plan` shows you.

```
ingress.4142503364.cidr_list.#:                "1" => "0"
ingress.4142503364.protocol:                   "ALL" => ""

ingress.727548830.cidr_list.#:                 "0" => "1"
ingress.727548830.protocol:                    "" => "ICMP"
```

> is it possible to separate the rules by a delimiter in the generated plan ?

The plan become unreadable, but that's the trade-off of using something less verbose.

**Comment by greut** [comment]
fab37f4eb623bf6b21b7d56b0fa5318722787a9ba084d9ab176c2edf2314e223  [terraform-provider-exo
scale_SNAPSHOT-2f1d05c71d45bc11d882eba0d30c42c3f26a056a_linux_amd64.tar.gz](https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_SNAPSHOT-2f1d05c71d45bc11d882eba0d30c42c3f26a056a_linux_amd64.tar.gz)
b5d8e97f90bc819b12eaf38747a88b5cf32998da355ae12e2760f1d04b5bdb26  [terraform-provider-exo
scale_SNAPSHOT-2f1d05c71d45bc11d882eba0d30c42c3f26a056a_darwin_amd64.tar.gz](https://sos-ch-dk-2.exo.io/yoanyoan/terraform-provider-exoscale_SNAPSHOT-2f1d05c71d45bc11d882eba0d30c42c3f26a056a_darwin_amd64.tar.gz)

**Pull request merged**